grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,165 @@
+
+  def __init__(self, config: dict) -> None:
+    """Constructor."""
+    self.url_map = Map([
+    Rule('/', endpoint='Root', methods=["GET"]),
+    Rule('/search', endpoint='Search', methods=["POST"]),
+    Rule('/query', endpoint='Query', methods=["POST"]),
+    Rule('/annotations', endpoint='Annotations', methods=["POST"])
+    ])
+
+  def dispatchRequest(self, request: JSONRequest) -> Any:
+    """Maps requests to different methods."""
+    adapter = self.url_map.bind_to_environ(request.environ)
+    try:
+      endpoint, values = adapter.match()
+      return getattr(self, 'on' + endpoint)(request, **values)
+    except HTTPException as e:
+      return e
+
+  def wsgiApp(self, environ, start_response) -> Response:
+    request = JSONRequest(environ)
+    response = self.dispatchRequest(request)
+    return response(environ, start_response)
+
+  def __call__(self, environ, start_response):
+    return self.wsgiApp(environ, start_response)
+
+  def onRoot(self, request: JSONRequest) -> JSONResponse:
+    """Returns OK message to database connection check."""
+    return JSONResponse()
+
+  def onSearch(self, request: JSONRequest) -> JSONResponse:
+    """Depending on the request type, returns either available client
+    resource usage metrics from Fleetspeak database, or possible values
+    for a defined Grafana variable (currently supports only variables based
+    on client IDs)."""
+    if "type" in request.json:
+      # Request issued on Panel > Queries page.
+      response = AVAILABLE_METRICS
+    else:
+      # Grafana issued request on Variables > New/Edit page.
+      response = fetchClientIds()  # todo: support Grafana variables other than ClientID.
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onQuery(self, request: JSONRequest) -> JSONResponse:
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    request_json_format = request.json
+    requested_client_id = extractClientIdFromVariable(request_json_format)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in request_json_format["targets"]]
+    response = fetchDatapointsForTargets(requested_client_id, request_json_format["maxDataPoints"], requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+def fetchClientIds() -> List[Text]:
+  """Fetches GRR client IDs that have resource usage records in Fleetspeak database."""
+  return fleetspeak_utils.GetClientIdsFromFleetspeak()
+
+
+Datapoint = Tuple[float, int]
+Datapoints = List[Datapoint]
+TargetWithDatapoints = Dict[Text, Datapoints]
+def fetchDatapointsForTargets(client_id: Text, limit: int, targets: List[Text]) -> List[TargetWithDatapoints]:
```

Comment:
If this function is not used outside of this module, please declare it as a module-private function: `_FetchDatapointsForTargets`.